### ISSUE TITLE
Consider moving `RemoveHandler` to return value of `ListenHandler`

### ISSUE THREAD

**Issue by keithamus** [rolftimmermans/event-iterator#8]
## Problem

`RemoveHandler` is useful for cleaning up event listeners that have been registered, but providing a `RemoveHandler` as the _second argument_ provides some limitations. 

Consider the following:

```ts
// Only stream right click events
const rightClicks = new EventIterator(push => {
  document.body.addEventListener('click', event => {
    if (event.button === 2) push(e)
  })
})
```

The way to clean up this event listener is to let a variable in the upper scope, capture it in the listener, and then reference that variable in the cleanup step:

```ts
// Only stream right click events
let listener = null
const rightClicks = new EventIterator(push => {
  listener = event => {
    if (event.button === 2) push(e)
  }
  document.body.addEventListener('click', listener)
}, () => {
  document.body.removeEventListener('click', listener)
})
```

This is problematic because calling `rightClicks[Symbol.asyncIterator()]` multiple times will override `listener` with new values.

## Potential Solution

One potential solution to avoid this is to allow the `ListenHandler` return a function of `RemoveHandler`:

```ts
type ListenHandler<T> = (
  push: PushCallback<T>,
  stop: StopCallback<T>,
  fail: FailCallback<T>
) => (() => void)|void
```

This way a `RemoveHandler` can be assigned to each invocation of `[Symbol.asyncIterator()]` with the lexical scope already implicit. The above `addEventListener` example now becomes:

```ts
// Only stream right click events
const rightClicks = new EventIterator(push => {
  const listener = event => {
    if (event.button === 2) push(e)
  }
  document.body.addEventListener('click', listener)
  return () => {
    document.body.removeEventListener('click', listener)
  }
})
```

The additional benefit to this remove handler is that it doesn't need to be given the `push/stop/fail` arguments as they're also implicitly in the lexical scope.

I think this would be a potentially breaking change - because the existing behavior of the return function is undefined. As it is a breaking change it might be useful to lean in and remove the second argument to `EventIterator` entirely, as having both would complicate the API.

Happy to raise a PR with code+tests+docs, but thought I'd raise the issue first to see if it has traction. Thoughts @rolftimmermans?

**Comment by rolftimmermans** [comment]
This is similar to React's `useEffect` so I guess some people might be familiar with the pattern. My initial reaction is very positive. It's definitely a breaking change though, but we can just push out a 2.0. Maybe with trivial compatibility wrapper for easy upgrading.

Apart from compatibility, are there any downsides to this?

**Comment by keithamus** [comment]
It's also similar to [Observable](https://github.com/tc39/proposal-observable) which has a very similar signature.

One downside to this is that if the `ListenHandler` call `stop` or `fail` synchronously, then the `RemoveHandler` won't be assigned and so cannot call. One way around the early `stop` or `fail` could be to wait one tick before executing those functions - simply wrapping the function body in a `Promise.resolve().then(...)`.

Throwing in the `ListenHandler` would also not call `RemoveHandler`, but that is the same as todays behaviour. However with this change `ListenHandler` throwing would mean `RemoveHandler` would never be assigned.

**Comment by rolftimmermans** [comment]
> One way around the early stop or fail could be to wait one tick before executing those functions - simply wrapping the function body in a Promise.resolve().then(...).

Yes, this sounds reasonable. But it should probably consistent. So the contract could be that calling `stop()` or `fail()` will never invoke the remove handler synchronously. 

I don't have an opinion about what throwing from the `ListenHandler` should mean. I guess if it does not call the remove handler now and it stays that way that should be okay.

Rethinking the API some more, I can think of another change we could consider (and combine with your idea, if this seems appealing to you as well). It might be nice to combine the 3 arguments into a single object, maybe called `EventQueue` (placeholder name).

```
interface EventQueue<T> {
  push(val: T): void
  stop(): void
  fail(err: Error): void
}

type RemoveHandler = () => void
type ListenHandler<T> = (events: EventQueue<T>) => void | RemoveHandler

new EventIterator<T>(listen: ListenHandler<T>)
```

This means you don't have to remember the order of `push`/`stop`/`fail` and can write:

```
new EventIterator(events => {
  this.addListener("data", events.push)
  this.addListener("end", events.stop)
  this.addListener("error", events.fail)

  return () => {
    this.removeListener("data", events.push)
    this.removeListener("end", events.stop)
    this.removeListener("error", events.fail)
    this.destroy()
  }
})
```

But this only seems an improvement if you need all handlers, not necessarily if you only need `push`.

**Comment by keithamus** [comment]
I'd be very happy to see one options bag argument instead of the three functions. For most of the instances I invoke I need one of either `push` or `push+fail`, I rarely use `stop` and so would be happy to just destructure like `new EventIterator({push, fail} => ...`

Shall I go ahead and raise a PR for this then?

**Comment by rolftimmermans** [comment]
> Shall I go ahead and raise a PR for this then?

That would be wonderful!

**Comment by Sceat** [comment]
Should this be closed now ?

**Comment by keithamus** [comment]
Yep!

**Comment by Sceat** [comment]
@keithamus can this be published ? :smile: #16

**Comment by keithamus** [comment]
I don't have publish rights. I think you meant to @ @rolftimmermans who is the author & maintainer of this project.
